training/deepspeech_training/util/signal_augmentations.py

```diff
@@ -0,0 +1,200 @@
+        sample.change_audio_type(new_audio_type=AUDIO_TYPE_NP)
+        rate = pick_value_from_range(self.rate, clock=clock)
+        audio = sample.audio
+        orig_len = len(audio)
+        audio = np.swapaxes(audio, 0, 1)
+        audio = resample(audio, sample.audio_format.rate, rate)
+        audio = resample(audio, rate, sample.audio_format.rate)
+        audio = np.swapaxes(audio, 0, 1)[0:orig_len]
+        sample.audio = audio
+
+
+class Codec:
+    """See "Codec augmentation" in TRAINING.rst"""
+    def __init__(self, p=1.0, bitrate=3200):
+        self.probability = float(p)
+        self.bitrate = int_range(bitrate)
+
+    def apply(self, sample, clock):
+        bitrate = pick_value_from_range(self.bitrate, clock=clock)
+        sample.change_audio_type(new_audio_type=AUDIO_TYPE_PCM)  # decoding to ensure it has to get encoded again
+        sample.change_audio_type(new_audio_type=AUDIO_TYPE_OPUS, bitrate=bitrate)  # will get decoded again downstream
+
+
+class Gaps:
+    """See "Gaps augmentation" in TRAINING.rst"""
+    def __init__(self, p=1.0, n=1, size=50.0):
+        self.probability = float(p)
+        self.n_gaps = int_range(n)
+        self.size = float_range(size)
+
+    def apply(self, sample, clock):
+        sample.change_audio_type(new_audio_type=AUDIO_TYPE_NP)
+        audio = sample.audio
+        n_gaps = pick_value_from_range(self.n_gaps, clock=clock)
+        for _ in range(n_gaps):
+            size = pick_value_from_range(self.size, clock=clock)
+            size = int(size * sample.audio_format.rate / 1000.0)
+            size = min(size, len(audio) // 10)  # a gap should never exceed 10 percent of the audio
+            offset = random.randint(0, max(0, len(audio) - size - 1))
+            audio[offset:offset + size] = 0
+        sample.audio = audio
+
+
+class Volume:
+    """See "Volume augmentation" in TRAINING.rst"""
+    def __init__(self, p=1.0, dbfs=3.0103):
+        self.probability = float(p)
+        self.target_dbfs = float_range(dbfs)
+
+    def apply(self, sample, clock):
+        sample.change_audio_type(new_audio_type=AUDIO_TYPE_NP)
+        target_dbfs = pick_value_from_range(self.target_dbfs, clock=clock)
+        sample.audio = normalize_audio(sample.audio, dbfs=target_dbfs)
+
+
+def parse_augmentation(augmentation_spec):
+    """
+    Parses an augmentation specification.
+
+    Parameters
+    ----------
+    augmentation_spec : str
+        Augmentation specification like "reverb[delay=20.0,decay=-20]".
+
+    Returns
+    -------
+    Instance of an augmentation class from util.signal_augmentations.*.
+    """
+    match = SPEC_PARSER.match(augmentation_spec)
+    if not match:
+        raise ValueError('Augmentation specification has wrong format')
+    cls_name = match.group(1)[0].upper() + match.group(1)[1:]
+    if cls_name not in globals():
+        raise ValueError('Unknown augmentation: {}'.format(cls_name))
```

Comment:
Maybe add an explicit registry of augmentations instead of using `globals()`?

Reply:
Code now checks if found class is sub-class of new `Augmentation` base class that I introduced for the `call_if_exists` removal.